### PR TITLE
SE - Non-nullable value types: Add support for constrained generics

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6105

This PR adds support for T in `<T> where T: struct` being recognized as non-nullable value type. 

Next PR: Tests for such constraints when used e.g. so `M(T? t)` will ensure that this is recognized as a nullable value type.

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/TypeHelperTest.cs [hunk 1]
```diff
@@ -89,5 +89,173 @@ public void Type_GetSymbolType_Alias()
...
+                }
+                """);
+            var field = tree.GetRoot().DescendantNodes().OfType<FieldDeclarationSyntax>().First();
+            var fieldSymbol = (IFieldSymbol)model.GetDeclaredSymbol(field.Declaration.Variables[0]);
+            var fieldType = fieldSymbol.Type;
+            fieldType.IsStruct().Should().BeFalse();
+        }
+
+        [DataTestMethod]
+        [DataRow("")]                      // Unbounded (can be nullable reference type or nullable value type)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/TypeHelperTest.cs:198]:
```suggestion
        [DataRow("")]                      // Unbounded (can be reference type or value type)
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/TypeHelperTest.cs [hunk 2]
```diff
@@ -89,5 +89,173 @@ public void Type_GetSymbolType_Alias()
...
+                struct CustomStruct { }
+                ref struct CustomRefStruct { }
+
+                ref struct Test
+                {
+                    {{type}} field;
+                }
+                """);
+            var field = tree.GetRoot().DescendantNodes().OfType<FieldDeclarationSyntax>().First();
+            var fieldSymbol = (IFieldSymbol)model.GetDeclaredSymbol(field.Declaration.Variables[0]);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/TypeHelperTest.cs:112]: Let's extract the `TestHelper.CompileCS` + finding syntax + finding `IFieldSymbol` into a method. Something like `IFieldSymbol CreateFieldSymbol(string code)`

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM, just some minor polishing

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 3]
```diff
@@ -532,15 +532,24 @@ public void Branching_LearnsObjectConstraint_RecursivePattern_ElseIsUnknown(stri
...
-            validator.ValidateTag("Else", x => x.Should().BeNull());
-            validator.ValidateTag("End", x => x.Should().BeNull());
+            validator.ValidateTag("If", x => x.Should().HaveNoConstraints());
+            validator.ValidateTag("Else", x => x.Should().HaveNoConstraints());
+            validator.ValidateTag("End", x => x.Should().HaveNoConstraints());
+        }
+
+        [DataTestMethod]
+        [DataRow("arg is { }", "TStruct")]
+        public void Branching_LearnsObjectConstraint_RecursivePattern_ValueTypeConstraint(string expression, string argType)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:547]: I added one more test case

**Pull request merged**